### PR TITLE
implemented baseline of sorts to do VQA

### PR DESCRIPTION
**Pull request by shashank2000**
Things I could definitely make better:

- Preprocessing of data - it's kinda messy right now
- Logic for dealing with variable length questions - this would require padding. Majority of the questions are 5/6 words long though.
- BatchNorm/Dropout wasn't added anywhere, I was just trying to get a base version running
- There are some stray comments in places for sure
- More logic for the answer embeddings - perhaps an RNN
- Lightning data module for train/val dataloaders, this will be in the next revision

### CODE REVIEW

File: project/data_preprocess.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from torchvision import transforms
```
Comment (mhw32) [inline project/data_preprocess.py]: What is this script for? is it just a test script?

Comment (mhw32) [reply project/data_preprocess.py]: (if so, might be good to not include in repo)

File: project/dataset/jeopardy_dataset.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import json
+from PIL import Image
+from torch.utils.data import Dataset
+import os
+import string
+from tqdm import tqdm
+
+class JeopardyDataset(Dataset):
+    """Face Landmarks dataset."""
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: Face landmarks?

File: project/dataset/jeopardy_dataset.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        self.images_dir = images_dir
+        self.image_id_to_filename = self._find_images()
+        self.transform = transform
+        self.question_to_answer = self._find_answers()
+        self.mega_dict = {} # this will store our indices
+        self.numerical = {}
+
+        answer_text = ' . '.join(list(self.question_to_answer.values()))
+        question_text = ' . '.join([q['question'] for q in self.questions])
+        question_text = question_text.replace("?", " END")
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: Why replace ? with END?

File: project/dataset/jeopardy_dataset.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        self.answers = json.load(a_f)["annotations"]
+        self.images_dir = images_dir
+        self.image_id_to_filename = self._find_images()
+        self.transform = transform
+        self.question_to_answer = self._find_answers()
+        self.mega_dict = {} # this will store our indices
+        self.numerical = {}
+
+        answer_text = ' . '.join(list(self.question_to_answer.values()))
+        question_text = ' . '.join([q['question'] for q in self.questions])
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: is this in case the question / answer are multiple sentences?

File: project/dataset/jeopardy_dataset.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+from PIL import Image
+from torch.utils.data import Dataset
+import os
+import string
+from tqdm import tqdm
+
+class JeopardyDataset(Dataset):
+    """Face Landmarks dataset."""
+
+    def __init__(self, questions_file, answers_file, images_dir, transform):
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: you probably need to support a 'train' split and a 'test' split

File: project/dataset/jeopardy_dataset.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        self.numerical = {}
+
+        answer_text = ' . '.join(list(self.question_to_answer.values()))
+        question_text = ' . '.join([q['question'] for q in self.questions])
+        question_text = question_text.replace("?", " END")
+        # strip all punctuation
+        question_text = question_text.translate(str.maketrans('', '', string.punctuation))
+        tokens = question_text.split(' ')
+        tokens.extend(answer_text.split(' '))
+        vocab = set(tokens)
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: when u add train/test splitting, make sure that you only make the vocab in the training split and reuse it in the test split

File: project/dataset/jeopardy_dataset.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        vocab_length = len(vocab)
+        print(vocab_length)
+        print("vocab length")
+        vocab = list(vocab)
+        word2idx = {w:i for i,w in enumerate(vocab)}
+        word2idx["DUMMY"] = len(vocab)
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: ?

File: project/dataset/jeopardy_dataset.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        
+        index = 0
+        for question in tqdm(self.questions):
+          arr = question['question']
+          arr = arr.replace("?", " END").translate(str.maketrans('', '', string.punctuation)).split()
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: write a helper function to do all the preprocessing of a sentence, otherwise as this gets more complicated, it could get very messy

File: project/dataset/jeopardy_dataset.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+          # dataset reduced by only 50%, and RNN is easier, don't have to worry about padding now
+          if len(arr) < 6 or len(arr) > 7:
+            continue
+          # add padding to arr, and separate the question mark
+          if len(arr) == 6:
+            arr.append("DUMMY")
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: we usually have three special tokens:

a start of sentence token `<s>`
an end of sentence token `</s>`
a padding token `<pad>`
and an unknown token `<unk>`

These need to be added into your vocabulary and u should build a sentence as follows:
```
<s> token1 token2 ... <pad> <pad> ... </s> 
```
where you pad to the appropriate length

File: project/dataset/jeopardy_dataset.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+          # len is always 7
+          if question["image_id"] in self.image_id_to_filename and question['question_id'] in self.question_to_answer:
+              self.numerical[question["question_id"]] = [word2idx[word] for word in arr]
+              self.mega_dict[index] = question, self.image_id_to_filename[question["image_id"]], word2idx[self.question_to_answer[question["question_id"]]]
+              if index < 10:
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: what is this for?

File: project/dataset/jeopardy_dataset.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        print("length is " + str(index))
+
+    def _find_answers(self):
+        # TODO: deal with the complications - there several answers for each question, 
+        # here's entry 0: {"answer": "net", "answer_confidence": "maybe", "answer_id": 1}
+        question_to_answer = {}
+        for ann in self.answers:
+            for answer in ann['answers']:
+              actual_ans = answer['answer']
+              if answer['answer_confidence'] == 'yes' and len(actual_ans.split()) == 1:
```
Comment (mhw32) [inline project/dataset/jeopardy_dataset.py]: why do we do this check?

Comment (mhw32) [reply project/dataset/jeopardy_dataset.py]: also, do we not need to convert the answer to a numerical?

File: project/model.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+from torch.optim import SGD
+import pytorch_lightning as pl 
+from torchvision.models import resnet18
+from loss_func.simclr import SimCLR
+import torch
+
+class JeopardyModel(pl.LightningModule):
+    def __init__(self, vocab_sz, bs):
+      super().__init__()
+      im_vec_dim = 128
```
Comment (mhw32) [inline project/model.py]: these should be cusotmiable hyperparameters that are in the init

File: project/model.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
+from torch.nn import Linear, Tanh, LSTM, Embedding
+from torch.optim import SGD
+import pytorch_lightning as pl 
+from torchvision.models import resnet18
+from loss_func.simclr import SimCLR
+import torch
+
+class JeopardyModel(pl.LightningModule):
+    def __init__(self, vocab_sz, bs):
```
Comment (mhw32) [inline project/model.py]: don't put bs as an argument, we generaly want to refrain from hardcoding to a bs

File: project/model.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+      
+      self.i_h = Embedding(vocab_sz, n_hidden)
+      self.rnn = LSTM(n_hidden, n_hidden)
+      self.h_o = Linear(n_hidden, question_dim)
+
+      self.h = None 
+      self.ans_final = Linear(n_hidden, ans_dim)
+
+      # for images, we use resnet18 pretrained
+      self.image_feature_extractor = resnet18(pretrained=True, progress=True)
```
Comment (mhw32) [inline project/model.py]: pretrained should not be set to true, we are learning frmo scratch. don't set progress to true either

File: project/model.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+      for p in self.image_feature_extractor.parameters():
```
Comment (mhw32) [inline project/model.py]: why are we cutting gradeints to these features? they wont learn otherwise

File: project/model.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+      self.image_feature_extractor.fc = Linear(512, im_vec_dim)
+      
+    def forward(self, x):
+      return self.image_feature_extractor(x)
+
+    def forward_question(self, x):
+      # we have a question as input
+      if not self.h:
+        self.h = torch.zeros(1, 16, 100, device=self.device), torch.zeros(1, 16, 100, device=self.device) # h0, c0
```
Comment (mhw32) [inline project/model.py]: where do these numbers 16 and 100 come from?

File: project/model.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+      res, h = self.rnn(self.i_h(x), self.h)
+      self.h = h[0].detach(), h[1].detach() # don't want to backpropagate all the way through the RNN
```
Comment (mhw32) [inline project/model.py]: ? don't you want to backprop through the rnn?

File: project/model.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+    def shared_step(self, batch):
+      question, image, answer = batch
+      question = torch.stack(question)
+      f_q = self.forward_question(question)
+      f_q = f_q.squeeze()[-1, :]
+      f_a = self.forward_answer(answer)
+      im_vector = self(image)
+      answer_image_vector = torch.cat((f_a, im_vector), 1)
+      loss = SimCLR(answer_image_vector, f_q).get_loss()
```
Comment (mhw32) [inline project/model.py:73]: awesome!

File: project/model.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+        self.h = torch.zeros(1, 16, 100, device=self.device), torch.zeros(1, 16, 100, device=self.device) # h0, c0
+      res, h = self.rnn(self.i_h(x), self.h)
+      self.h = h[0].detach(), h[1].detach() # don't want to backpropagate all the way through the RNN
+      return self.h_o(res)
+
+    def forward_answer(self, x):
+      # just a linear layer over the embeddings to begin with
+      x = self.i_h(x)
+      return self.ans_final(x)
+
```
Comment (mhw32) [inline project/model.py:49]: extra space

File: project/data_preprocess.py (hunk 1 above)
Comment (mhw32) [reply project/data_preprocess.py]: oh i see how you are using this from later one, can you wrap this code into your lightning system?

File: project/data_preprocess.py [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+train_indices, val_indices = indices[split:], indices[:split]
+
+# Creating PT data samplers and loaders:
+train_sampler = SubsetRandomSampler(train_indices)
+valid_sampler = SubsetRandomSampler(val_indices)
+
+train_loader = DataLoader(dataset, batch_size=batch_size, 
+                                           sampler=train_sampler, num_workers=4, pin_memory=True, drop_last=True)
+val_loader = DataLoader(dataset, batch_size=batch_size,
```
Comment (mhw32) [inline project/data_preprocess.py]: it is kind of cheating to make the vocab on both the training and validation split. I suggest you fix a split of training and test and then define the vocab only on the training set. For unseen tokens in the validation, replace with <unk>

File: project/run.py [hunk 21]
```diff
@@ -0,0 +1,12 @@
+from data_preprocess import train_loader, val_loader, batch_size
+from PIL import ImageFile
+import pytorch_lightning as pl
+from model import JeopardyModel
+
+ImageFile.LOAD_TRUNCATED_IMAGES = True
+
+# add gpus=1 if training on a GPU; progress_bar_refresh_rate=50 if on Jupyter notebook
+trainer = pl.Trainer(max_epochs=10, progress_bar_refresh_rate=1)
```
Comment (mhw32) [inline project/run.py]: we will want to train for 200 epochs when gpu comes

**Review by mhw32 — CHANGES_REQUESTED** [review]

File: project/dataset/jeopardy_dataset.py (hunk 2 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: removed

File: project/dataset/jeopardy_dataset.py (hunk 3 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: So I was a little bit unsure about the choices surrounding this - I'd ideally just strip out all punctuation, but doesn't the question mark have special semantic meaning since we're dealing with questions half the time. Perhaps we can talk about this? I strip away all punctuation in the sentence later on, and I guess I wanted to "save" the question mark from this

File: project/dataset/jeopardy_dataset.py (hunk 9 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: Gotcha - made the relevant changes.

File: project/dataset/jeopardy_dataset.py (hunk 7 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: fixed with your padding note

File: project/dataset/jeopardy_dataset.py (hunk 10 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: this was just a sanity check for me to make sure there were some indices getting inserted

File: project/model.py (hunk 15 above)
Comment (shashank2000) [reply project/model.py]: froze weights because I didn't want to wait forever for the training loop; removed this stuff now that I'll be using cluster

File: project/model.py (hunk 16 above)
Comment (shashank2000) [reply project/model.py]: 16 is the batch size, 100 is n_hidden (should have just made that an instance variable).

File: project/model.py (hunk 19 above)
Comment (shashank2000) [reply project/model.py:49]: fixed

File: project/model.py (hunk 13 above)
Comment (shashank2000) [reply project/model.py]: so the only place I use the batch size is when I initialize the LSTM hidden state and cell state with all zeroes - and that's why I need this argument. I'll think of possible workarounds

File: project/data_preprocess.py (hunk 20 above)
Comment (shashank2000) [reply project/data_preprocess.py]: this makes a lot of sense - what would I do for unseen tokens in the validation? Replace with -1 or something?

File: project/dataset/jeopardy_dataset.py (hunk 5 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: so I guess the idea would be I'd get rid of the random sampler stuff for now at least, and just make it so that any datapoint above index 0.8*len(dataset) is in the test set? And these questions words won't be added to the vocab

File: project/dataset/jeopardy_dataset.py (hunk 4 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: I'm not sure I understand the question - end goal is to tokenize, perhaps you mean I could have done it with just 
`' '.join()`? That's what I do in the revision

File: project/dataset/jeopardy_dataset.py (hunk 11 above)
Comment (shashank2000) [reply project/dataset/jeopardy_dataset.py]: I do this line 64. The vocab isn't formed yet at this point.

File: project/model.py (hunk 17 above)
Comment (shashank2000) [reply project/model.py]: I was under the impression this was done to prevent exploding/vanishing gradients? I was looking at this [post](https://discuss.pytorch.org/t/solved-why-we-need-to-detach-variable-which-contains-hidden-representation/1426/11) for context. Took off detach for now, perhaps we can talk about it?

File: project/model.py [hunk 22]
```diff
@@ -75,4 +78,4 @@ def shared_step(self, batch):
       return loss
 
     def configure_optimizers(self):
-      return SGD(self.parameters(), lr=2e-2)
+      return Adam(self.parameters(), lr=2e-2)
```
Comment (mhw32) [inline project/model.py:77]: Quick a high learning rate, use SGD with momentum 0.9, learning rate 0.03 and weight decay 1e-4
